cmd/controller/main.go

```diff
@@ -234,19 +242,99 @@ func main() {
 		rs = append(rs, gasController)
 	}
 
-	kubeInformerFactory.Start(ctx.Done())
-	agonesInformerFactory.Start(ctx.Done())
+	// Start of the leader election code
+	id := uuid.New().String()
 
-	for _, r := range rs {
-		go func(rr runner) {
-			if runErr := rr.Run(ctx, ctlConf.NumWorkers); runErr != nil {
-				logger.WithError(runErr).Fatalf("could not start runner: %T", rr)
-			}
-		}(r)
+	klog.InitFlags(nil)
+	flag.Set("v", fmt.Sprintf("%d", 5))
+	klog.V(4).Info("Checking if V4s get printd")
+
+	config, err := buildConfig(ctlConf.KubeConfig)
+	if err != nil {
+		logger.Fatal(err)
+	}
+	client := kubernetes.NewForConfigOrDie(config)
+
+	run := func(ctx context.Context) {
+		kubeInformerFactory.Start(ctx.Done())
+		agonesInformerFactory.Start(ctx.Done())
+
+		for _, r := range rs {
+			go func(rr runner) {
+				if runErr := rr.Run(ctx, ctlConf.NumWorkers); runErr != nil {
+					logger.WithError(runErr).Fatalf("could not start runner: %T", rr)
+				}
+			}(r)
+		}
 	}
 
-	<-ctx.Done()
-	logger.Info("Shut down agones controllers")
+	lock := &resourcelock.LeaseLock{
+		LeaseMeta: metav1.ObjectMeta{
+			Name:      "agones-controller",
+			Namespace: "agones-system",
+		},
+		Client: client.CoordinationV1(),
+		LockConfig: resourcelock.ResourceLockConfig{
+			Identity: id,
+		},
+	}
+
+	logger.Infof("This controller ID is %s", id)
+
+	leaderelection.RunOrDie(ctx, leaderelection.LeaderElectionConfig{
+		Lock: lock,
+		// IMPORTANT: you MUST ensure that any code you have that
+		// is protected by the lease must terminate **before**
+		// you call cancel. Otherwise, you could have a background
+		// loop still running and another process could
+		// get elected before your background loop finished, violating
+		// the stated goal of the lease.
+		ReleaseOnCancel: true,
+		LeaseDuration:   60 * time.Second,
+		RenewDeadline:   15 * time.Second,
+		RetryPeriod:     5 * time.Second,
+		Callbacks: leaderelection.LeaderCallbacks{
+			OnStartedLeading: func(ctx context.Context) {
+				// we're notified when we start - this is where you would
+				// usually put your code
+				logger.Infof("Leader elected: %s", id)
+				run(ctx)
+
+				<-ctx.Done()
+				logger.Info("Shut down agones controllers")
```

Comment:
I would move these two things inside `run()` and maybe just rename it to literally `runWhenLeader` for clarity.

Actually for readability I might take this whole `RunOrDie` and move it to a helper function so that the code in main() looks something like this:

```
onLeaderElection(func () {
  kubeInformerFactory.Start(ctx.Done())
  agonesInformerFactory.Start(ctx.Done())
  [...]
})
```

That way if you're just trying to understand the flow of main(), you see the inline function and it's easy to realize "oh, after leader election it's going to run <this>". And then if you care about the details of leader election, you can review the helper function.